### ISSUE TITLE
Synchronously calling next() will run cb() synchronously

### ISSUE THREAD

**Issue by ArtskydJ** [paulmillr/async-each#8]
http://blog.izs.me/post/59142742143/designing-apis-for-asynchrony

```js
console.log('should be 1')

function sync(a, next) {
    next(null, a)
}

each([ 'alpha', 'bravo', 'charlie' ], sync, function (err, result) {
    console.log('should be 3')
})

console.log('should be 2')
```

```
should be 1
should be 3
should be 2
```

The callback should always be called asynchronously.

**Comment by es128** [comment]
This module does not *enforce* asynchrony, it is compatible with it. It is consistent in its handling, and therefore does not inherently cause zalgo issues - although it certainly is possible as you've demonstrated for users to do so with it.

It's worth noting that https://github.com/caolan/async behaves exactly the same with its `each` method.

I spent a few minutes experimenting with this and prepared an implementation that does enforce async like you're requesting, but I've also confirmed it would be a major breaking change if incorporated. And I don't think that's what we want to do. I'm pushing it [in a branch](https://github.com/paulmillr/async-each/tree/enforce-async) in case it's interesting to you.
